server/project_member.go

```diff
@@ -2,18 +2,161 @@ package server
+				}
+
+				gitlabProjectMemberList, err := vcsPlugin.Get("GITLAB_SELF_HOST", vcsPlugin.ProviderConfig{Logger: s.l}).FetchProjectMemberList(ctx,
+					common.OauthContext{
+						ClientID:     vcs.ApplicationID,
+						ClientSecret: vcs.Secret,
+						AccessToken:  repo.AccessToken,
+						RefreshToken: repo.RefreshToken,
+						Refresher:    s.refreshToken(ctx, repo.ID),
+					},
+					vcs.InstanceURL,
+					repo.ExternalID,
+				)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member from GitLab, instance URL: %s", vcs.InstanceURL)).SetInternal(err)
+				}
+
+				// create or patch project member
+				findProjectMember := &api.ProjectMemberFind{ProjectID: &projectID}
+				bytebaseProjectMemberList, err := s.ProjectMemberService.FindProjectMemberList(ctx, findProjectMember)
+				if err != nil {
+					return echo.NewHTTPError(http.StatusNotFound, fmt.Sprintf("fail to fetch project member in bytebase: %d", projectID)).SetInternal(err)
+				}
+
+				// check whether principal exists in our system. if not exist, create one.
+				for _, projectMember := range gitlabProjectMemberList {
+					findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+					principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+					if err != nil {
+						return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+					}
+
+					if projectMember.State != vcsPlugin.UserStateActive && projectMember.MembershipState != vcsPlugin.UserStateActive {
+						continue
+					}
+
+					if principal == nil {
+						signupInfo := &api.Signup{
+							Name:     projectMember.Name,
+							Email:    projectMember.Email,
+							Password: common.RandomString(20),
+						}
+						createdPrincipal, httpErr := TrySignup(ctx, s, signupInfo, api.PrincipalAuthProviderGitlabSelfHost, c.Get(getPrincipalIDContextKey()).(int))
+						if httpErr != nil {
+							return httpErr
+						}
+						principal = createdPrincipal
+					}
+
+					isProjectMemberExist := false
+					for _, bytebaseProjectMember := range bytebaseProjectMemberList {
+						if bytebaseProjectMember.PrincipalID == principal.ID {
+							payload, _ := json.Marshal(projectMember)
+							// try update status
+							patchProjectMember := &api.ProjectMemberPatch{
+								ID:           bytebaseProjectMember.ID,
+								RoleProvider: api.ProjectRoleProviderGitlabSelfHost,
+								Payload:      string(payload),
+							}
+							_, err := s.ProjectMemberService.PatchProjectMember(ctx, patchProjectMember)
+							if err != nil {
+								return echo.NewHTTPError(http.StatusInternalServerError, "Failed to sync member from GitLab").SetInternal(err)
+							}
+							isProjectMemberExist = true
+						}
+					}
+
+					if !isProjectMemberExist {
+						var role api.ProjectRole
+						switch projectMember.AccessLevel { // see https://docs.gitlab.com/ee/api/members.html
+						case 50, /* Owner */
```

Comment:
nit: the format is a bit off

Reply:
So it is ok that Bytebase Role is visible to VCS plugin?